### DISCUSSION TITLE
get siblings values

### DISCUSSION THREAD

**Discussion by diramazioni** [freeplane/freeplane#763] [Help] [answered]
I wonder if is possible to get the values of siblings attributes or nodes in formulas (or scripting if necessary) 
i.e like in a spreedsheet get the sum of all the attributes displayed for the node.
`node.parent.children.sum()` creates a circular dependency and node.siblings doesn't exist
Another question is how to get all the attributes regardless of their names?

**Comment by macmarrum** [comment, marked as answer]
You can use Groovy standard methods like `findAll` to exclude the calling node from the collection of children, and `collect` to extract specific fields.

To go over each attribute, you can do `node.attributes.collect { attr -> "${attr.key}: ${attr.value}" }`

```groovy
def siblings = node.parent.children.findAll { it != node }
def sumAttributeValues(_node) {
  _node.attributes.collect { attr -> _node[attr.key].num0 }.sum(0)
}
siblings.collect { sib -> sumAttributeValues(sib) }.sum()
```

**Comment by diramazioni** [comment]
Thanks, this is close to what I was looking for, but not exactly.
[esercizi1.mm.zip](https://github.com/freeplane/freeplane/files/9824160/esercizi1.mm.zip)

In this example your code correctly sum all the attributes of the siblings nodes (9 in the example). 
However I was looking for 
- a way to sum the attributes of the same node (like in the the first table last row)
- a way to sum (or make other operations) on the node siblings, 
I've tried to follow your example and it works (30) not sure it's the best way to write it, Say for example I want to compute the mean?

**Reply by macmarrum** [reply]
> a way to sum the attributes of the same node (like in the the first table last row)

The same principle applies here. First you can exclude the calling attribute ("siblings sum") using `findAll`, then you can extract the values using `collect`, finally sum it all up.
```groovy
=node.attributes.findAll { it.key != 'siblings sum' }.collect { node[it.key].num0 }.sum(0)
```

> a way to sum (or make other operations) on the node siblings,
I've tried to follow your example and it works (30) not sure it's the best way to write it

It looks good to me :+1:
You could make it a one-liner
```groovy
=node.parent.children.findAll { it != node }.collect { it.to.num0 }.sum(0)
```

> Say for example I want to compute the mean?

If all nodes contain numbers, you can use [average()](https://docs.groovy-lang.org/docs/groovy-4.0.1/html/groovy-jdk/java/lang/Iterable.html#average())
```groovy
=node.parent.children.findAll { it != node }.collect { it.to.num }.average()
```
In case there are non-numeric siblings, you'd need to define an algorithm that suits your needs.

___

Some interesting reading about the stuff in curly braces
- https://groovy-lang.org/closures.html

**Comment by diramazioni** [comment]
awesome, I think I start to understand a little the groovy syntax, I wonder if I can make it a function a call it among the different nodes, how you would do that?

**Reply by macmarrum** [reply]
See https://docs.freeplane.org/attic/old-mediawiki-content/Your_own_utility_script_library.html

In essence, create a file in `{user-dir}/lib/`, e.g. named `FP.groovy`.
_Note: You can use `Tools->Open user directory`_
Write your function(s) there, with the `static` keyword, e.g.
```groovy
def static sumUpOtherAttributes(node, attrName = 'siblings sum') {
    node.attributes.findAll { it.key != attrName }.collect { node[it.key].num0 }.sum(0)
}
```

Restart Freeplane.

In you attribute (siblings sum) put 
```groovy
=FP.sumUpOtherAttributes(node)
```
